File: src/docs/namecheap-hosting/local-client.py

import requests
import json
import time
import logging
from typing import Dict, List, Optional
from datetime import datetime
import os
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

class MEvApiClient:
    def __init__(self, relay_url: str, api_key: Optional[str] = None):
        self.relay_url = relay_url.rstrip('/')
        self.api_key = api_key
        self.session = requests.Session()
        
        if api_key:
            self.session.headers.update({'X-API-Key': api_key})
        
        # Setup logging
        logging.basicConfig(
            level=logging.INFO,
            format='%(asctime)s - %(levelname)s - %(message)s',
            handlers=[
                logging.FileHandler('mev_client.log'),
                logging.StreamHandler()
            ]
        )
        self.logger = logging.getLogger(__name__)

    def get_live_prices(self, tokens: List[str]) -> Dict:
        """Get live prices through the relay server"""
        try:
            url = f"{self.relay_url}/api/live-feed/prices"
            payload = {"tokens": tokens}
            
            response = self.session.post(url, json=payload, timeout=10)
            response.raise_for_status()
            
            data = response.json()
            
            cache_status = "CACHED" if data.get('cached') else "FRESH"
            self.logger.info(f"Prices fetched for {tokens} - {cache_status}")
            
            return data
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Failed to fetch prices: {e}")
            return {"success": False, "error": str(e)}

    def get_arbitrage_opportunities(self) -> Dict:
        """Get arbitrage opportunities"""
        try:
            url = f"{self.relay_url}/api/live-feed/arbitrage"
            response = self.session.get(url, timeout=10)
            response.raise_for_status()
            
            data = response.json()
            opportunities = data.get('opportunities', [])
            
            self.logger.info(f"Found {len(opportunities)} arbitrage opportunities")
            
            return data
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Failed to fetch opportunities: {e}")
            return {"success": False, "error": str(e)}

    def monitor_prices(self, tokens: List[str], interval: int = 5):
        """Continuously monitor prices"""
        self.logger.info(f"Starting price monitoring for {tokens} (interval: {interval}s)")
        
        while True:
            try:
                # Get prices
                prices = self.get_live_prices(tokens)
                if prices.get('success'):
                    for price_data in prices.get('data', []):
                        symbol = price_data['symbol']
                        price = price_data['price']
                        change = price_data['change24h']
                        
                        print(f"{symbol}: ${price:.4f} ({change:+.2f}%)")
                
                # Get opportunities
                opportunities = self.get_arbitrage_opportunities()
                if opportunities.get('success'):
                    for opp in opportunities.get('opportunities', [])[:3]:  # Top 3
                        profit = opp['profitOpportunity']
                        confidence = opp['confidence']
                        
                        print(f"📈 {opp['tokenA']}-{opp['tokenB']}: {profit:.2f}% profit ({confidence:.1f}% confidence)")
                
                print("-" * 50)
                time.sleep(interval)
                
            except KeyboardInterrupt:
                self.logger.info("Monitoring stopped by user")
                break
            except Exception as e:
                self.logger.error(f"Monitoring error: {e}")
                time.sleep(interval)

    def send_webhook_test(self, webhook_url: str, data: Dict):
        """Send test webhook"""
        try:
            response = requests.post(webhook_url, json=data, timeout=5)
            response.raise_for_status()
            
            self.logger.info("Webhook sent successfully")
            return response.json()
        except Exception as e:
            self.logger.error(f"Webhook failed: {e}")
            return {"success": False, "error": str(e)}

# Example usage and configuration
if __name__ == "__main__":
    # Configuration
    RELAY_URL = os.getenv('RELAY_URL', 'https://your-domain.com')
    API_KEY = os.getenv('API_KEY')
    
    # Initialize client
    client = MEvApiClient(RELAY_URL, API_KEY)
    
    # Test the connection
    health = requests.get(f"{RELAY_URL}/health").json()
    print(f"Relay server status: {health['status']}")
    
    # Monitor specific tokens
    tokens_to_monitor = ['SOL', 'ETH', 'USDC', 'USDT']
    
    try:
        client.monitor_prices(tokens_to_monitor, interval=5)
    except KeyboardInterrupt:
        print("\nMonitoring stopped.")

File: src/docs/namecheap-hosting/local-client.js

const axios = require('axios');
const EventEmitter = require('events');
require('dotenv').config();

class MEvApiClient extends EventEmitter {
  constructor(relayUrl, apiKey = null) {
    super();
    this.relayUrl = relayUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
    this.isMonitoring = false;
    
    this.client = axios.create({
      baseURL: this.relayUrl,
      timeout: 10000,
      headers: apiKey ? { 'X-API-Key': apiKey } : {}
    });
  }

  async getLivePrices(tokens) {
    try {
      const response = await this.client.post('/api/live-feed/prices', { tokens });
      
      const cacheStatus = response.data.cached ? 'CACHED' : 'FRESH';
      console.log(`✅ Prices fetched for ${tokens.join(', ')} - ${cacheStatus}`);
      
      return response.data;
    } catch (error) {
      console.error('❌ Failed to fetch prices:', error.message);
      return { success: false, error: error.message };
    }
  }

  async getArbitrageOpportunities() {
    try {
      const response = await this.client.get('/api/live-feed/arbitrage');
      
      const opportunities = response.data.opportunities || [];
      console.log(`🔍 Found ${opportunities.length} arbitrage opportunities`);
      
      return response.data;
    } catch (error) {
      console.error('❌ Failed to fetch opportunities:', error.message);
      return { success: false, error: error.message };
    }
  }

  async startMonitoring(tokens, interval = 5000) {
    if (this.isMonitoring) {
      console.log('⚠️ Monitoring already active');
      return;
    }

    this.isMonitoring = true;
    console.log(`🚀 Starting price monitoring for ${tokens.join(', ')} (interval: ${interval}ms)`);

    const monitor = async () => {
      if (!this.isMonitoring) return;

      try {
        // Fetch prices
        const prices = await this.getLivePrices(tokens);
        if (prices.success) {
          this.emit('prices', prices.data);
          
          prices.data.forEach(priceData => {
            const { symbol, price, change24h } = priceData;
            const changeColor = change24h >= 0 ? '🟢' : '🔴';
            console.log(`${changeColor} ${symbol}: $${price.toFixed(4)} (${change24h.toFixed(2)}%)`);
          });
        }

        // Fetch opportunities
        const opportunities = await this.getArbitrageOpportunities();
        if (opportunities.success) {
          this.emit('opportunities', opportunities.opportunities);
          
          opportunities.opportunities.slice(0, 3).forEach(opp => {
            const { tokenA, tokenB, profitOpportunity, confidence } = opp;
            console.log(`📈 ${tokenA}-${tokenB}: ${profitOpportunity.toFixed(2)}% profit (${confidence.toFixed(1)}% confidence)`);
          });
        }

        console.log('-'.repeat(50));
        
      } catch (error) {
        console.error('❌ Monitoring error:', error.message);
        this.emit('error', error);
      }

      setTimeout(monitor, interval);
    };

    monitor();
  }

  stopMonitoring() {
    this.isMonitoring = false;
    console.log('⏹️ Monitoring stopped');
  }

  async sendWebhook(webhookUrl, data) {
    try {
      const response = await axios.post(webhookUrl, data, { timeout: 5000 });
      console.log('📤 Webhook sent successfully');
      return response.data;
    } catch (error) {
      console.error('❌ Webhook failed:', error.message);
      return { success: false, error: error.message };
    }
  }

  async testConnection() {
    try {
      const response = await this.client.get('/health');
      console.log(`✅ Relay server healthy - Uptime: ${response.data.uptime}s`);
      return true;
    } catch (error) {
      console.error('❌ Connection test failed:', error.message);
      return false;
    }
  }
}

// Example usage
async function main() {
  const RELAY_URL = process.env.RELAY_URL || 'https://your-domain.com';
  const API_KEY = process.env.API_KEY;
  
  const client = new MEvApiClient(RELAY_URL, API_KEY);
  
  // Event listeners
  client.on('prices', (prices) => {
    // Handle price updates
    console.log('📊 Price update received');
  });
  
  client.on('opportunities', (opportunities) => {
    // Handle arbitrage opportunities
    const highProfitOpps = opportunities.filter(opp => opp.profitOpportunity > 1);
    if (highProfitOpps.length > 0) {
      console.log(`🚨 ${highProfitOpps.length} high-profit opportunities detected!`);
    }
  });
  
  client.on('error', (error) => {
    console.error('Client error:', error.message);
  });
  
  // Test connection
  const isConnected = await client.testConnection();
  if (!isConnected) {
    console.error('Failed to connect to relay server');
    process.exit(1);
  }
  
  // Start monitoring
  const tokensToMonitor = ['SOL', 'ETH', 'USDC', 'USDT'];
  await client.startMonitoring(tokensToMonitor, 5000);
  
  // Graceful shutdown
  process.on('SIGINT', () => {
    console.log('\n🛑 Shutting down...');
    client.stopMonitoring();
    process.exit(0);
  });
}

if (require.main === module) {
  main().catch(console.error);
}

module.exports = MEvApiClient;
